docs/source/etcd-data-model.rst

```diff
@@ -339,23 +339,49 @@ Each rule sub-object has the following JSON-encoded structure:
 .. code-block:: json
 
     {
-      "protocol": "tcp|udp|icmp|icmpv6",
+      # Positive matches:
+      "protocol": "tcp | udp | icmp | icmpv6",
       "src_tag": "<tag_name>",
+      "src_selector": "<selector expression>",
       "src_net": "<CIDR>",
       "src_ports": [1234, "2048:4000"],
       "dst_tag": "<tag_name>",
       "dst_net": "<CIDR>",
       "dst_ports": [1234, "2048:4000"],
-      "icmp_type": <int>,
-      "action": "deny|allow",
+      "icmp_type": <int>, "icmp_code": <int>,  # Treated together, see below.
+
+      # Negated matches:
+      "!protocol": ...,
+      "!src_tag": ...,
+      "!src_selector": ...,
```

Comment:
A selector with negation inside it will render the known (endpoint) IPs that don't match the selector.  "!src_selector" renders the positive ipset and then negates the match so it'll match IPs that are not assigned to endpoints too.